In [3]:
import pandas as pd

In [4]:
pd.DataFrame({'Yes':[50,21],'No':[131,2]})

Yes   No
0   50  131
1   21    2

In [5]:
pd.DataFrame({'Bob':['I Liked it.','It was awful.'], 'Sue':['Pretty good.','Bland.']})

Bob           Sue
0    I Liked it.  Pretty good.
1  It was awful.        Bland.

In [6]:
pd.DataFrame({'Bob':['I Liked it.','It was awful.'], 'Sue':['Pretty good.','Bland.']}, index=['Product A', 'Product B'])

Bob           Sue
Product A    I Liked it.  Pretty good.
Product B  It was awful.        Bland.

In [7]:
pd.Series([1, 2, 3, 4, 5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [8]:
pd.Series([30, 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')

2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: int64

In [9]:
pd.Series(['2015 Sales', '2016 Sales', '2017 Sales'], index=[30, 35, 40], name='Product A')

30    2015 Sales
35    2016 Sales
40    2017 Sales
Name: Product A, dtype: object

In [10]:
df = pd.read_csv("user_behavior_dataset.csv")

In [11]:
df

User ID        Device Model Operating System  App Usage Time (min/day)  \
0          1      Google Pixel 5          Android                       393   
1          2           OnePlus 9          Android                       268   
2          3        Xiaomi Mi 11          Android                       154   
3          4      Google Pixel 5          Android                       239   
4          5           iPhone 12              iOS                       187   
..       ...                 ...              ...                       ...   
695      696           iPhone 12              iOS                        92   
696      697        Xiaomi Mi 11          Android                       316   
697      698      Google Pixel 5          Android                        99   
698      699  Samsung Galaxy S21          Android                        62   
699      700           OnePlus 9          Android                       212   

     Screen On Time (hours/day)  Battery Drain (mAh/day)  \
0                           6.4                     1872   
1                           4.7                     1331   
2                           4.0                      761   
3                           4.8                     1676   
4                           4.3                     1367   
..                          ...                      ...   
695                         3.9                     1082   
696                         6.8                     1965   
697                         3.1                      942   
698                         1.7                      431   
699                         5.4                     1306   

     Number of Apps Installed  Data Usage (MB/day)  Age  Gender  \
0                          67                 1122   40    Male   
1                          42                  944   47  Female   
2                          32                  322   42    Male   
3                          56                  871   20    Male   
4                          58                  988   31  Female   
..                        ...                  ...  ...     ...   
695                        26                  381   22    Male   
696                        68                 1201   59    Male   
697                        22                  457   50  Female   
698                        13                  224   44    Male   
699                        49                  828   23  Female   

     User Behavior Class  
0                      4  
1                      3  
2                      2  
3                      3  
4                      3  
..                   ...  
695                    2  
696                    4  
697                    2  
698                    1  
699                    3  

[700 rows x 11 columns]

In [12]:
df.shape

(700, 11)

In [13]:
df.head()

User ID    Device Model Operating System  App Usage Time (min/day)  \
0        1  Google Pixel 5          Android                       393   
1        2       OnePlus 9          Android                       268   
2        3    Xiaomi Mi 11          Android                       154   
3        4  Google Pixel 5          Android                       239   
4        5       iPhone 12              iOS                       187   

   Screen On Time (hours/day)  Battery Drain (mAh/day)  \
0                         6.4                     1872   
1                         4.7                     1331   
2                         4.0                      761   
3                         4.8                     1676   
4                         4.3                     1367   

   Number of Apps Installed  Data Usage (MB/day)  Age  Gender  \
0                        67                 1122   40    Male   
1                        42                  944   47  Female   
2                        32                  322   42    Male   
3                        56                  871   20    Male   
4                        58                  988   31  Female   

   User Behavior Class  
0                    4  
1                    3  
2                    2  
3                    3  
4                    3

In [14]:
gym = pd.read_csv("gym.csv")
gym

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
0     56    Male         88.3        1.71      180      157           60   
1     46  Female         74.9        1.53      179      151           66   
2     32  Female         68.1        1.66      167      122           54   
3     25    Male         53.2        1.70      190      164           56   
4     38    Male         46.1        1.79      188      158           68   
..   ...     ...          ...         ...      ...      ...          ...   
968   24    Male         87.1        1.74      187      158           67   
969   25    Male         66.6        1.61      184      166           56   
970   59  Female         60.4        1.76      194      120           53   
971   32    Male        126.4        1.83      198      146           62   
972   46    Male         88.7        1.63      166      146           66   

     Session_Duration (hours)  Calories_Burned Workout_Type  Fat_Percentage  \
0                        1.69           1313.0         Yoga            12.6   
1                        1.30            883.0         HIIT            33.9   
2                        1.11            677.0       Cardio            33.4   
3                        0.59            532.0     Strength            28.8   
4                        0.64            556.0     Strength            29.2   
..                        ...              ...          ...             ...   
968                      1.57           1364.0     Strength            10.0   
969                      1.38           1260.0     Strength            25.0   
970                      1.72            929.0       Cardio            18.8   
971                      1.10            883.0         HIIT            28.2   
972                      0.75            542.0     Strength            28.8   

     Water_Intake (liters)  Workout_Frequency (days/week)  Experience_Level  \
0                      3.5                              4                 3   
1                      2.1                              4                 2   
2                      2.3                              4                 2   
3                      2.1                              3                 1   
4                      2.8                              3                 1   
..                     ...                            ...               ...   
968                    3.5                              4                 3   
969                    3.0                              2                 1   
970                    2.7                              5                 3   
971                    2.1                              3                 2   
972                    3.5                              2                 1   

       BMI  
0    30.20  
1    32.00  
2    24.71  
3    18.41  
4    14.39  
..     ...  
968  28.77  
969  25.69  
970  19.50  
971  37.74  
972  33.38  

[973 rows x 15 columns]

In [15]:
gym.BMI.mean()

24.91212744090442

In [16]:
gym.shape

(973, 15)

# Indexing, Selecting & Assigning

In [17]:
gym.Workout_Type

0          Yoga
1          HIIT
2        Cardio
3      Strength
4      Strength
         ...   
968    Strength
969    Strength
970      Cardio
971        HIIT
972    Strength
Name: Workout_Type, Length: 973, dtype: object

In [18]:
gym['Workout_Type'][69]

'HIIT'

In [19]:
gym.iloc[0]

Age                                  56
Gender                             Male
Weight (kg)                        88.3
Height (m)                         1.71
Max_BPM                             180
Avg_BPM                             157
Resting_BPM                          60
Session_Duration (hours)           1.69
Calories_Burned                  1313.0
Workout_Type                       Yoga
Fat_Percentage                     12.6
Water_Intake (liters)               3.5
Workout_Frequency (days/week)         4
Experience_Level                      3
BMI                                30.2
Name: 0, dtype: object

In [20]:
gym.iloc[:, 0]

0      56
1      46
2      32
3      25
4      38
       ..
968    24
969    25
970    59
971    32
972    46
Name: Age, Length: 973, dtype: int64

In [21]:
gym.iloc[:, 7]

0      1.69
1      1.30
2      1.11
3      0.59
4      0.64
       ... 
968    1.57
969    1.38
970    1.72
971    1.10
972    0.75
Name: Session_Duration (hours), Length: 973, dtype: float64

In [22]:
gym.iloc[:, 7].mean()

1.2564234326824255

In [23]:
gym.iloc[:4, 7]

0    1.69
1    1.30
2    1.11
3    0.59
Name: Session_Duration (hours), dtype: float64

In [24]:
gym.iloc[3:12, 7]

3     0.59
4     0.64
5     1.59
6     1.49
7     1.27
8     1.03
9     1.08
10    0.82
11    1.15
Name: Session_Duration (hours), dtype: float64

In [25]:
gym.iloc[3:12, 7].mean()

1.0733333333333333

In [26]:
gym.iloc[3:9, 0]

3    25
4    38
5    56
6    36
7    40
8    28
Name: Age, dtype: int64

In [27]:
gym.iloc[[3,4,5], 0]

3    25
4    38
5    56
Name: Age, dtype: int64

In [28]:
gym.iloc[-5:]

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
968   24    Male         87.1        1.74      187      158           67   
969   25    Male         66.6        1.61      184      166           56   
970   59  Female         60.4        1.76      194      120           53   
971   32    Male        126.4        1.83      198      146           62   
972   46    Male         88.7        1.63      166      146           66   

     Session_Duration (hours)  Calories_Burned Workout_Type  Fat_Percentage  \
968                      1.57           1364.0     Strength            10.0   
969                      1.38           1260.0     Strength            25.0   
970                      1.72            929.0       Cardio            18.8   
971                      1.10            883.0         HIIT            28.2   
972                      0.75            542.0     Strength            28.8   

     Water_Intake (liters)  Workout_Frequency (days/week)  Experience_Level  \
968                    3.5                              4                 3   
969                    3.0                              2                 1   
970                    2.7                              5                 3   
971                    2.1                              3                 2   
972                    3.5                              2                 1   

       BMI  
968  28.77  
969  25.69  
970  19.50  
971  37.74  
972  33.38

In [29]:
gym

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
0     56    Male         88.3        1.71      180      157           60   
1     46  Female         74.9        1.53      179      151           66   
2     32  Female         68.1        1.66      167      122           54   
3     25    Male         53.2        1.70      190      164           56   
4     38    Male         46.1        1.79      188      158           68   
..   ...     ...          ...         ...      ...      ...          ...   
968   24    Male         87.1        1.74      187      158           67   
969   25    Male         66.6        1.61      184      166           56   
970   59  Female         60.4        1.76      194      120           53   
971   32    Male        126.4        1.83      198      146           62   
972   46    Male         88.7        1.63      166      146           66   

     Session_Duration (hours)  Calories_Burned Workout_Type  Fat_Percentage  \
0                        1.69           1313.0         Yoga            12.6   
1                        1.30            883.0         HIIT            33.9   
2                        1.11            677.0       Cardio            33.4   
3                        0.59            532.0     Strength            28.8   
4                        0.64            556.0     Strength            29.2   
..                        ...              ...          ...             ...   
968                      1.57           1364.0     Strength            10.0   
969                      1.38           1260.0     Strength            25.0   
970                      1.72            929.0       Cardio            18.8   
971                      1.10            883.0         HIIT            28.2   
972                      0.75            542.0     Strength            28.8   

     Water_Intake (liters)  Workout_Frequency (days/week)  Experience_Level  \
0                      3.5                              4                 3   
1                      2.1                              4                 2   
2                      2.3                              4                 2   
3                      2.1                              3                 1   
4                      2.8                              3                 1   
..                     ...                            ...               ...   
968                    3.5                              4                 3   
969                    3.0                              2                 1   
970                    2.7                              5                 3   
971                    2.1                              3                 2   
972                    3.5                              2                 1   

       BMI  
0    30.20  
1    32.00  
2    24.71  
3    18.41  
4    14.39  
..     ...  
968  28.77  
969  25.69  
970  19.50  
971  37.74  
972  33.38  

[973 rows x 15 columns]

In [30]:
gym.loc[0, 'Weight (kg)']

88.3

In [31]:
gym.loc[0, ['Weight (kg)', 'Max_BPM', 'Calories_Burned']]

Weight (kg)          88.3
Max_BPM               180
Calories_Burned    1313.0
Name: 0, dtype: object

In [32]:
gym.loc[:, ['Weight (kg)', 'Height (m)', 'Fat_Percentage']]

Weight (kg)  Height (m)  Fat_Percentage
0           88.3        1.71            12.6
1           74.9        1.53            33.9
2           68.1        1.66            33.4
3           53.2        1.70            28.8
4           46.1        1.79            29.2
..           ...         ...             ...
968         87.1        1.74            10.0
969         66.6        1.61            25.0
970         60.4        1.76            18.8
971        126.4        1.83            28.2
972         88.7        1.63            28.8

[973 rows x 3 columns]

# Manipulating Index

In [33]:
gym.set_index("Weight (kg)")

Age  Gender  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
Weight (kg)                                                           
88.3          56    Male        1.71      180      157           60   
74.9          46  Female        1.53      179      151           66   
68.1          32  Female        1.66      167      122           54   
53.2          25    Male        1.70      190      164           56   
46.1          38    Male        1.79      188      158           68   
...          ...     ...         ...      ...      ...          ...   
87.1          24    Male        1.74      187      158           67   
66.6          25    Male        1.61      184      166           56   
60.4          59  Female        1.76      194      120           53   
126.4         32    Male        1.83      198      146           62   
88.7          46    Male        1.63      166      146           66   

             Session_Duration (hours)  Calories_Burned Workout_Type  \
Weight (kg)                                                           
88.3                             1.69           1313.0         Yoga   
74.9                             1.30            883.0         HIIT   
68.1                             1.11            677.0       Cardio   
53.2                             0.59            532.0     Strength   
46.1                             0.64            556.0     Strength   
...                               ...              ...          ...   
87.1                             1.57           1364.0     Strength   
66.6                             1.38           1260.0     Strength   
60.4                             1.72            929.0       Cardio   
126.4                            1.10            883.0         HIIT   
88.7                             0.75            542.0     Strength   

             Fat_Percentage  Water_Intake (liters)  \
Weight (kg)                                          
88.3                   12.6                    3.5   
74.9                   33.9                    2.1   
68.1                   33.4                    2.3   
53.2                   28.8                    2.1   
46.1                   29.2                    2.8   
...                     ...                    ...   
87.1                   10.0                    3.5   
66.6                   25.0                    3.0   
60.4                   18.8                    2.7   
126.4                  28.2                    2.1   
88.7                   28.8                    3.5   

             Workout_Frequency (days/week)  Experience_Level    BMI  
Weight (kg)                                                          
88.3                                     4                 3  30.20  
74.9                                     4                 2  32.00  
68.1                                     4                 2  24.71  
53.2                                     3                 1  18.41  
46.1                                     3                 1  14.39  
...                                    ...               ...    ...  
87.1                                     4                 3  28.77  
66.6                                     2                 1  25.69  
60.4                                     5                 3  19.50  
126.4                                    3                 2  37.74  
88.7                                     2                 1  33.38  

[973 rows x 14 columns]

In [36]:
gym.BMI<=25 

0      False
1      False
2       True
3       True
4       True
       ...  
968    False
969    False
970     True
971    False
972    False
Name: BMI, Length: 973, dtype: bool

In [37]:
gym.loc[gym.Workout_Type=='Yoga']

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
0     56    Male         88.3        1.71      180      157           60   
14    20    Male        117.7        1.81      196      161           54   
22    50  Female         71.0        1.68      187      161           70   
23    29    Male        120.9        1.78      197      168           65   
25    42  Female         63.7        1.71      173      169           62   
..   ...     ...          ...         ...      ...      ...          ...   
950   31  Female         76.7        1.62      174      127           74   
953   50  Female         67.9        1.77      171      164           73   
961   31  Female         66.6        1.67      184      137           71   
965   23  Female         44.1        1.62      196      122           58   
967   20    Male         55.0        1.60      172      168           67   

     Session_Duration (hours)  Calories_Burned Workout_Type  Fat_Percentage  \
0                        1.69           1313.0         Yoga            12.6   
14                       1.35           1195.0         Yoga            26.4   
22                       1.17            848.0         Yoga            33.1   
23                       0.78            721.0         Yoga            28.1   
25                       1.42           1080.0         Yoga            26.1   
..                        ...              ...          ...             ...   
950                      1.39            883.0         Yoga            28.1   
953                      1.03            760.0         Yoga            33.0   
961                      1.02            699.0         Yoga            31.0   
965                      0.58            354.0         Yoga            25.7   
967                      1.12           1035.0         Yoga            24.0   

     Water_Intake (liters)  Workout_Frequency (days/week)  Experience_Level  \
0                      3.5                              4                 3   
14                     3.3                              3                 2   
22                     2.5                              2                 1   
23                     3.4                              2                 1   
25                     1.8                              3                 2   
..                     ...                            ...               ...   
950                    2.3                              4                 2   
953                    2.4                              3                 2   
961                    1.5                              4                 2   
965                    2.7                              2                 1   
967                    3.2                              4                 2   

       BMI  
0    30.20  
14   35.93  
22   25.16  
23   38.16  
25   21.78  
..     ...  
950  29.23  
953  21.67  
961  23.88  
965  16.80  
967  21.48  

[239 rows x 15 columns]

That means among 973 people 239 do Yoga

In [39]:
gym.loc[(gym.Workout_Type=='Yoga') & (gym.BMI<=25)]

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
25    42  Female         63.7        1.71      173      169           62   
46    24    Male         54.7        1.74      160      167           53   
50    34  Female         51.0        1.62      162      158           53   
61    51  Female         44.8        1.63      198      137           69   
68    31  Female         61.1        1.76      189      150           67   
..   ...     ...          ...         ...      ...      ...          ...   
921   29  Female         75.6        1.80      175      149           57   
953   50  Female         67.9        1.77      171      164           73   
961   31  Female         66.6        1.67      184      137           71   
965   23  Female         44.1        1.62      196      122           58   
967   20    Male         55.0        1.60      172      168           67   

     Session_Duration (hours)  Calories_Burned Workout_Type  Fat_Percentage  \
25                       1.42           1080.0         Yoga            26.1   
46                       0.84            772.0         Yoga            22.9   
50                       1.42           1122.0         Yoga            30.8   
61                       1.08            666.0         Yoga            33.8   
68                       1.81           1358.0         Yoga            18.6   
..                        ...              ...          ...             ...   
921                      1.31            976.0         Yoga            25.7   
953                      1.03            760.0         Yoga            33.0   
961                      1.02            699.0         Yoga            31.0   
965                      0.58            354.0         Yoga            25.7   
967                      1.12           1035.0         Yoga            24.0   

     Water_Intake (liters)  Workout_Frequency (days/week)  Experience_Level  \
25                     1.8                              3                 2   
46                     3.6                              3                 1   
50                     2.4                              3                 2   
61                     2.1                              4                 2   
68                     2.7                              4                 3   
..                     ...                            ...               ...   
921                    1.9                              4                 2   
953                    2.4                              3                 2   
961                    1.5                              4                 2   
965                    2.7                              2                 1   
967                    3.2                              4                 2   

       BMI  
25   21.78  
46   18.07  
50   19.43  
61   16.86  
68   19.72  
..     ...  
921  23.33  
953  21.67  
961  23.88  
965  16.80  
967  21.48  

[135 rows x 15 columns]

Among 239 people who do Yoga, 135 people has BMI less or equal to 25 which is standard BMI level

In [40]:
gym.loc[(gym.Workout_Type=='Yoga') & (gym.BMI<=25) & (gym.BMI>=20)]

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
25    42  Female         63.7        1.71      173      169           62   
82    24  Female         72.9        1.76      175      127           68   
84    41    Male         72.1        1.83      175      133           58   
102   29    Male         58.6        1.61      198      122           72   
115   54    Male         67.0        1.68      191      164           63   
..   ...     ...          ...         ...      ...      ...          ...   
915   48    Male         73.1        1.83      194      131           71   
921   29  Female         75.6        1.80      175      149           57   
953   50  Female         67.9        1.77      171      164           73   
961   31  Female         66.6        1.67      184      137           71   
967   20    Male         55.0        1.60      172      168           67   

     Session_Duration (hours)  Calories_Burned Workout_Type  Fat_Percentage  \
25                       1.42           1080.0         Yoga            26.1   
82                       0.86            546.0         Yoga            30.8   
84                       1.13            744.0         Yoga            21.4   
102                      1.20            805.0         Yoga            23.4   
115                      0.55            446.0         Yoga            20.5   
..                        ...              ...          ...             ...   
915                      1.05            681.0         Yoga            28.0   
921                      1.31            976.0         Yoga            25.7   
953                      1.03            760.0         Yoga            33.0   
961                      1.02            699.0         Yoga            31.0   
967                      1.12           1035.0         Yoga            24.0   

     Water_Intake (liters)  Workout_Frequency (days/week)  Experience_Level  \
25                     1.8                              3                 2   
82                     1.5                              3                 1   
84                     2.8                              3                 2   
102                    2.0                              3                 2   
115                    2.8                              3                 1   
..                     ...                            ...               ...   
915                    2.8                              4                 2   
921                    1.9                              4                 2   
953                    2.4                              3                 2   
961                    1.5                              4                 2   
967                    3.2                              4                 2   

       BMI  
25   21.78  
82   23.53  
84   21.53  
102  22.61  
115  23.74  
..     ...  
915  21.83  
921  23.33  
953  21.67  
961  23.88  
967  21.48  

[68 rows x 15 columns]

68 people who does Yoga has a standard BMI (20<=BMI<=25) 